Add database functionality

TomDijkema opened the pull request:
Adds database functionality to the application.

Add:
- Function for inserting records into the countries or organisations table
- Function for requesting countries or organisations and reformatting data

Modify:
- Graph functions now rely on the database functionality instead of CSV
- CSV functions are now suppressed but still remain in the repository

TomDijkema committed:
Add database functionality

Adds database functionality to the application.

Add:
- Function for inserting records into the countries or organisations table
- Function for requesting countries or organisations and reformatting data

Modify:
- Graph functions now rely on the database functionality instead of CSV
- CSV functions are now suppressed but still remain in the repository

TomDijkema commented:
I will look at the complexity of functions, figure out a way to simplify them

TomDijkema committed:
Refactored code

Commit refactors code to be less complex and pass the SonarLint analysis. Functions that draw graphs have been split up in multiple ones that are called from a specific directory: /draw_functions.

Adds:
- New /draw_functions directory with draw graph specific functions
- Option to select all countries, not bound to request list

Modifies:
- graphical_functions file to be less complex, parts of functions have been simplified or moved to /draw_functions as seperate files

TomDijkema committed:
Refactor Code #2

Second commit for refactoring code in behalf of SonarCloud analysis

samleeflang commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+# Copy the content of the local src directory to the working directory
+COPY . .
+
+# Adding file permission for csv files
```
Is this still necessary?

TomDijkema replied:
Removed CSV references in Docker file, however, Docker file does not properly start python script, could you please have a look at the changes of the Dockerfile in the new commit?

samleeflang commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+# Set the working directory in the container
+WORKDIR /code
+
+# Set csv_files directory as mounting volume
```
Maybe just remove them if it is no longer necessary

TomDijkema replied:
Removed CSV references in Docker file, however, Docker file does not properly start python script, could you please have a look at the changes of the Dockerfile in the new commit?

samleeflang commented on query_database.py:
```diff
@@ -0,0 +1,405 @@
+import psycopg2
```
https://www.sqlalchemy.org/ probably makes a couple of things easier.
It is also database agnostic so if we want to switch to something is it is a bit more generic

TomDijkema replied:
Replaced core psycopg2 usage with SQLAlcehmy, psycopg2 is still being used as the database driver for Postgresql and is now included in the Dockerfile which needed the special package libpg-dev and a C compiler (build-essential)

samleeflang commented on kubernetes.yaml:
```diff
@@ -0,0 +1,20 @@
+        spec:
+          containers:
+            - name: network-status-overview
+              image: network-status-overview
```
I think image should be the full path to the AWS image repo. Image is probably not pushed to that yet.

TomDijkema replied:
Image has not been pushed yet indeed, I shall insert a reference in the file for remembrance, probably good to first have a look at the Dockerfile before uploading the image to AWS

samleeflang commented on main2.py:
```diff
@@ -0,0 +1,42 @@
+
+
+# Import database logic
+# import query_database
```
Remove unused import

TomDijkema replied:
Done

samleeflang commented on main2.py:
```diff
@@ -0,0 +1,42 @@
+
+
+def process_countries():
+    logging.info('\nReceiving countries datasets data from GBIF...')
```
The '\n' should be necessary to enforce a new line. Are you sure this is needed now you use the logger?

TomDijkema replied:
Correct assumption, removed the '\n' elements

samleeflang commented on package-lock.json:
```diff
@@ -0,0 +1,11 @@
+{
```
What does this file do?

TomDijkema replied:
The ifle is important for the fixed range of versioning packages and is automatically generated by npm: 
https://nodejs.dev/learn/the-package-lock-json-file

samleeflang replied:
Ok thanks.

samleeflang commented on main2.py:
```diff
@@ -0,0 +1,42 @@
+    query_database.insert_organisations_data(gbif_organisations_data, geocase_data)
+
+
+process_countries()
```
Isn't a main function necessary here. If this becomes the primary main.py it is probably needed to start the scripts?

TomDijkema replied:
Added main function, will be optimized in upcoming version when application will be split up into two smaller components

samleeflang commented on query_database.py:
```diff
@@ -0,0 +1,405 @@
+        conn = psycopg2.connect(**params)
+        cur = conn.cursor()
+
+        # Check if entity already exists (for same month)
```
This is also possible with an upsert query:
https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-upsert/

TomDijkema replied:
Combined separate insert and update functions (queries) to one upset query for both ways

samleeflang commented on query_database.py:
```diff
@@ -0,0 +1,405 @@
+            # Insert new record
+            query = insert_countries(country_entity)
+
+        # Execute query
```
See if you can use the try-with-resources functionality then you don't need to explicitly commit and close the connection.
For sql alchemy:
https://docs.sqlalchemy.org/en/14/core/connections.html

TomDijkema replied:
Replaced plain Psycopg2 database queries with SQL Alchemy queries that use the 'with' method

samleeflang reviewed (commented):
Nice that the database works and this looks good.
Check if you can remove the csv references some more so we don't leave unused code in there.
For db management switching to sqlalchemy would probably be good

TomDijkema committed:
Review commit 1

Adds:
- SQLAlchemy to make quering the database easier and more structured
- Main function for main2.py
- Models that represent the countries and organisations tables

Modifies:
- Dockerfile to remove CSV file references and incorporate new packages

TomDijkema committed:
Merge branch 'main' into feature-graph_endpoint
